designs/default-zero-value.md

```diff
@@ -0,0 +1,778 @@
+}
+```
+
+
+## Deprecations in Smithy IDL 1.0 and removals in IDL 2.0
+
+To simplify member nullability, the `@box` trait will be deprecated in Smithy
+1.0 for structure members, and removed entirely in Smithy 2.0. Instead of
+relying on the `@box` trait, Smithy validation will encourage modelers to
+target nullable shapes rather than make a non-nullable shape null. This
+validation will make the transition for models from Smithy 1.x to Smithy 2.x
+easier because the removal of the `@box` trait will not impact a model. The
+Smithy Java implementation's `NullableIndex` will account for the new and
+existing nullability rules.
+
+The following shapes in the prelude will be marked as deprecated in Smithy 1.0,
+and removed in Smithy 2.0: `PrimitiveBoolean`, `PrimitiveShort`,
+`PrimitiveInteger`, `PrimitiveLong`, `PrimitiveFloat`, and `PrimitiveDouble`.
+Instead of referring to these shapes, refer to their non-primitive shape
+variants (for example, `Integer`, `Boolean`) and apply the `@default` trait to
+structure members. A warning will be emitted when models refer to these shapes,
+which encourages models to update to use shapes that are forward compatible
+with an eventual Smithy 2.0.
+
+Smithy 2.0 IDL support will be added to the 1.x series of Smithy's Java tooling
+so that both 1.0 and 2.0 models can be loaded, simultaneously, and automatically
+converted from 1.0 to 2.0.
+
+
+## New guidance for code generators
+
+Code generated types for structures SHOULD use the `@default` trait and
+`@required` traits to provide member accessors that always return non-null
+values.
+
+- When the `@default` trait is present, the corresponding accessor always
+  returns a non-null value by defaulting missing members with their zero values.
+- When the `@required` trait is present on a member that targets a shape that
+  has a zero value, the corresponding accessor always returns a non-null value.
+- Smithy implementations in languages like TypeScript that do not provide a kind
+  of constructor or builder to create structures may not be able to set default
+  values, precluding them from being able to treat `@required` and `@default`
+  members as non-null.
+
+
+### Serialization
+
+Members marked with the `@default` trait that are set to their default zero
+values SHOULD NOT be serialized.
+
+- Deserializers MUST tolerate receiving a serialized default value. Failing here
+  would leave end-users with no recourse other than to wait for a client or
+  server to update and fix their implementation. This also accounts for older
+  clients that think a structure member is required, but the service has since
+  transitioned the member to use the `@default` trait.
+- Serializing zero values can cause unintended information disclosure like
+  revealing internal-only members (this is the same reason that null values
+  SHOULD NOT be serialized).
+
+
+### Deserialization
+
+Structure members marked as `@required` are not in and of themselves a guarantee
```

Comment:
I have a lot of thoughts on this PR and writing them all will take a little bit of time that I'm hoping to find next week. 

However, the implications of this paragraph constitutes a significant risk for how we're designing the use of smithy in my organisation, and I find myself compelled to write something now. I completely understand that the project is primarily shaped by AWS usecases, but I think this statement (as well as a lot of details in your write-up) ought to be protocol-specific, at the very least. Otherwise you're essentially forcing all protocols to be akin to protobuf v3, which is somewhat controversial and unpractical in many cases. 

But, further than that, what constitutes a breaking change is extremely dependant on whether the types are used in covariant (output) or contravariant (input) position. It is important to recognise that, and I'll happily share a write-up I did on the subject, next monday. 

I 100% believe that having clients have to accept the absence of value on @required fields during deserialisation is wrong. When a service goes from @required to @default on an output type, my opinion is that the service should have to send the default value to not break existing clients, and for clients to protect against what is essentially breaking behaviour from services.

Reply:
Thanks for taking a look and leaving your thoughts. Here are a few high level points that I can work back into the doc:

1. The goal of this doc isn't to dictate protocols, but rather how we think about nullabiltiy and backward compatibility around nullability in Smithy, and how that impacts codegen. If code generators are compliant with this proposal, then they won't get broken by services that think they're making backward compatible changes. If we are making claims that too broadly impact protocol design, we'll adjust the proposal.
2. In Smithy, protocols don't change how types are code-generated. The only thing that should change between protocols is how types are serialized and deserialized. Changing the protocol a client uses to call a service should not change code-generated types, which means what shapes do and don't have a zero value needs to be standardized in a protocol agnostic way. That's a kind of North Star for features in Smithy, which is what we'll need to arrive at with this proposal.
3. We should be able to use the same code-generated types for both client and server codegen use cases, and we also have use cases for Smithy where types are used outside of client/server interactions.
4. The default trait isn't only for transitioning something from required -> optional. It's used to provide a default value for members that don't need nullability, like when 0 and unset are equivalent, defaulting to 0 makes a service easier to use. You can actually do this today in Smithy with the `smithy.api#PrimitiveInteger` shape and other shapes in the prelude. This proposal is expanding that to other types.

> what constitutes a breaking change is extremely dependant on whether the types are used in covariant (output) or contravariant (input) position

We did consider this, and I can work this into the FAQ. For example, making an input member transition from required -> optional might not break a client because it's less restrictive, and neither would making an output member transition from optional -> required because output types are typically created only during deserialization (though things like mocking and injecting responses are cases where this isn't true).

But the thing about this proposal is that it's about the types we code generate from Smithy models. Using required traits and default traits to influence code generation means that the changes I just listed would be breaking changes to many Smithy code generators because the types would change from `T` -> `Option<T>` or `Option<T>` -> `T`. While this only will impact code generators that can bake required traits into their type systems, specifications around what is and isn't backward compatible in models gives these code generators the affordances they need in order to generate more idiomatic types.

I think one way to support different nullability rules for input and output is to bifurcate codegen so that shapes used as part of input are generated separately from shapes used as part of output. There are a lot of tradeoffs to this though, for example code size increase (2 generated types per shape) and awkwardness of roundtripping output shapes as input. 

More intrinsically, a design goal of Smithy is that we should be able to use the same code-generated types for both client and server codegen (we're building Smithy service codegen as we speak!). We also have use cases for Smithy where types are used outside of client/server interactions, which requires codegen guidance to be based on types and not if they're used as part of input or output shapes.

> I 100% believe that having clients have to accept the absence of value on `@required` fields during deserialisation is wrong. When a service goes from `@required` to `@default` on an output type, my opinion is that the service should have to send the default value to not break existing clients, and for clients to protect against what is essentially breaking behaviour from services.

I agree this could be up to a protocol to decide (for example, maybe you need to reuse a protocol that doesn't know about Smithy's default value semantics). I can update this section to more generally talk about how different protocols might choose to serialize or not serialize default values. Keep in mind though that the default trait isn't only for transitioning something from required -> optional; for these kinds of use cases, always serializing a default value could lead to information disclosure (e.g., a new member that hasn't yet been made available to all users yet), which is actually why we typically don't serialize null values in Smithy.

I'm curious why this matters for your use case though -- if a required member is missing during deserialization, you can fill in a deterministic and constant zero value (that's the part where clients are protecting themselves from breaking changes, right?).

Reply:
Thanks for the detailed answer. I'll start with the points I agree with : 

### Things with which I agree

> In Smithy, protocols don't change how types are code-generated 

> We should be able to use the same code-generated types for both client and server codegen use cases, and we also have use cases for Smithy where types are used outside of client/server interactions.

I very much agree with these points, and the code generator I'm developing abide by these rules. This is good design and greatly facilitates testability and conceptual understanding of what constitutes a service. It also greatly reduces bikeshedding discussions. That is one think I very much like with smithy : the fact that protocol semantics are separated from core semantics in the IDL guides the design of tooling to a state that unlocks a number of amazing possibilities, when protocol-level details more often pollute the generated code  

> More intrinsically, a design goal of Smithy is that we should be able to use the same code-generated types for both client and server codegen (we're building Smithy service codegen as we speak!)

Once again, I'm very much in agreement that this is good design. I have built both server/client side support using the same generated code. My generated code isn't even specialised based on whether the usecase is client or server, I just have a bunch of abstractions that let me interpret code in client-side or server-side context, it works beautifully and my users are very happy with it. 

>   what shapes do and don't have a zero value needs to be standardized in a protocol agnostic way. 

I can certainly agree with that. 

### What I do not agree with : Smithy dictating what code generators ought to do in the ecosystem. 

So one thing that I am not in agreement with is that the smithy specification (as found in https://awslabs.github.io/smithy/) dictates rules about the generated code. 

Let's take the example of timestamps, as seen in [this table] : 

> The serialization of a timestamp is an implementation detail that is determined by a protocol and MUST NOT have any effect on the types exposed by tooling to represent a timestamp value. 

The main thing that I disagree with in that statement is the use of the `MUST` word (as opposed to `SHOULD`). I think the documentation has to offer an educated advice on what works and what doesn't. I for one think that AWS has a lot more experience in this field than any other organisation in the world, and I very much agree with this particular advice regarding how timestamps are to be translated to generated code, but any third party building smithy tooling is completely empowered to make that decision for themselves, as long as the data that transits over the wire is compliant to the semantics of every protocol it integrates with. With that in mind, the use of `MUST` is not only impossible to test/enforce at the scale of the ecosystem (which the smithy developers have no control over), but also feels un-necessarily prescriptive and I'm willing to bet it will raise eyebrows (some of the spec has certainly surprised me). 

> But the thing about this proposal is that it's about the types we code generate from Smithy models

Going back to the concept of "null", some languages (like haskell for instance, but a great many other) do not even support it. The absence of value is always modelled as an algebraic `Option` or `Maybe` datatype. Having the smithy spec state how a `@default` trait `must` translate in languages that do not support nulls make it seem like the smithy spec "overextends" into topics the authors of the specification might not be fully versed in. It might be that haskell folks think would prefer for @default to translate to `Maybe`, as it may be more idiomatic to these languages. Whether this choice presents a risk around library compatibility when specs change should be (imho) the responsibility of the person maintaining code generators for this language, certainly not of the smithy-spec authors. 

Changing `MUST` by `SHOULD` would remove the impression that the smithy-spec authors assume to know best for every possible language. Rather, it means they offer a (very) educated opinion on what works well in the languages that they have experience with, which is a totally valuable thing to do.  

So, though I totally think it's the prerogative of the smithy-team (or AWS) to be dictating rules about the code generators they maintain themselves, I do however think it's unrealistic for these rules to be worded as "sacrosanct" in smithy's specification.  

#### Going back to the topic of protocols 

> I agree this could be up to a protocol to decide (for example, maybe you need to reuse a protocol that doesn't know about Smithy's default value semantics). I can update this section to more generally talk about how different protocols might choose to serialize or not serialize default values.

That would be greatly appreciated. 

> Keep in mind though that the default trait isn't only for transitioning something from required -> optional; 

I have very much kept this in mind throughout this discussion. However, the semantics of `optional` (as in the absence of value) and `default` are easy to conflate, at the protocol level.  In my organisation, services and clients are reasonably often on a footing, and the data that transits over the wire is the result of discussions between the two. Whatever is `@required` is always meant to be filled with a meaningful value, and for smithy to prescribe that a `@required` => `@default` modification is always non-breaking, whilst dictating that it's okay for @default fields to sometimes not appear over the wire, would mean that it's okay for services to start breaking the contractual promise to always fill a meaningful value there. 

Smithy dictating this (regardless of protocols) also remove organisations ability to define protocols where the absence of value and the presence of a default value (in the serialised messages) have the same meaning. This might not occur often, and you might think that the difference is meaningless in most usecases you think about, but I certainly think this should be discussed at the level of protocols. 

> for these kinds of use cases, always serializing a default value could lead to information disclosure (e.g., a new member that hasn't yet been made available to all users yet), which is actually why we typically don't serialize null values in Smithy.

I think if that actually something people worry about, then neither @default nor @optional are the right modelling. But I 100% think that it's AWS prerogative to use it as such in the AWS protocols 😄 . 

Also I kinda dislike the wording "we typically don't serialize null values in Smithy", when I think you mean "we typically don't serialise null values in the AWS-specific protocols we defined using Smithy". I think the distinction is important to be made if you want smithy to outlive the boundaries of AWS and become an industry standard ... which I very much think it should ! 

#### Summary : 

* For reasons I've already expressed, I very much agree with a plan to removal of `@box` and the shapes prefixed with  `Primitive`, because they are obviously tied to some language-level concerns (in particular JVM), which (imho) inadvertently signals some undesirable coupling between smithy and programming languages. 
* I think `@default` is a sensible path forward
* I think the smithy specification should not dictate what code generators do, but rather give educated opinion on the matter (ie replace MUST by SHOULD)
* I think whether `0` values transit on the wire when a shape is annotated with `@default` should be a protocol-level decision rather than a IDL-level decision (in a similar fashion to how protocols are able to state whether they support `Documents` or not). 

#### Post scriptum : 

I hope that this discussion is not perceived as a lack of appreciation or distrust. I think the `smithy` language and core libraries (`smithy-model`, `smithy-diff`, etc) are already showing immense value, and I'm really grateful for this product.

Reply:
Regarding timestamps and codegen: I understand your broader point about codegen, but I want to address your specific example about timestamps. I think the current guidance is necessary because it sets expectations of what modelers can expect from timestamp shapes. Unambiguous codegen is a goal of Smithy, and these explicit statements about what affordances the metamodel and traits provide is important (even if some requirements can't be validated through compliance checks). If this timestamp requirement became a "SHOULD" instead of a "MUST", then there's ambiguity as to what implementations need to support when generating timestamp types. The "MUST" reinforces that the timestamp shape is part of Smithy's metamodel and has specific meaning as a type in the metamodel (that is, "an instant in time with no UTC offset or timezone").

If an implementation needs to deviate from the spec, then they can, for example, use a string shape with a custom trait to indicate that the string contains a timestamp in a specific format and with specific semantics (like a timezone or something). They can then add codegen extensions to the code generators they use to honor the trait.

> The absence of value is always modelled as an algebraic Option or Maybe datatype. Having the smithy spec state how a `@default` trait must translate in languages that do not support nulls make it seem like the smithy spec "overextends" into topics the authors of the specification might not be fully versed in.

This proposal was created in response to AWS building code generators in languages like Rust that do not have explicit nulls, but rather use `Option<T>` and `T` (like `Maybe` in Haskell). These kinds of languages were a key consideration when writing this proposal. We're confident that implementors for Haskell, Rust, and other programming languages are used to mapping null concepts from other systems like JSON Schema to their languages, and we didn't think the use of the word null would cause undue confusion. I mention in the motivation section that nullability is equivalent to optionality, but would a more tactical solution here be to move that guidance into the intro?

> It might be that haskell folks think would prefer for `@default` to translate to Maybe, as it may be more idiomatic to these languages.

Using `Maybe` for members marked with the `@default` trait would only make sense to me if an implementation just treats all members as optional (some of the generators we're building in AWS will probably function this way due to limitations in different PLs).

However, I think explicitly using `Maybe` for members marked with the `@default` trait would be a misunderstanding of the default trait, and it would make that implementation use features that are not afforded to it by Smithy's metamodel (which makes the implementation at risk of incompatibility with other Smithy implementations). The `@default` trait does not imply a tri-state of set, unset, or default. The usage of the `@default` trait forfeits the knowledge of whether a type is set explicitly to a zero value or defaulted to a zero value. If that information is required, then `@default` / `@required` shouldn't be used. For this use case you need to use nullable/optional/maybe/whatever you wanna call it members, which provides the member the ability to differentiate between set and unset.

>  Whether this choice presents a risk around library compatibility when specs change should be (imho) the responsibility of the person maintaining code generators for this language, certainly not of the smithy-spec authors.
> ...
> Changing MUST by SHOULD would remove the impression that the smithy-spec authors assume to know best for every possible language. Rather, it means they offer a (very) educated opinion on what works well in the languages that they have experience with, which is a totally valuable thing to do.

I think this is a major benefit of Smithy. By standardizing on backward compatibility guarantees and API lifecycle changes over time, we make implementations more consistent with each other and we easily diff models for backward compatibility changes. It becomes easier to reason about models generally in API reviews rather than needing to understand every possible and trait and protocol used by the model.

Ultimately, the decision does lie with the implementor, but the specification is setting expectations around what every other implementation expects. Deviating from the spec tells implementors that they're signing up for way more work because they need to customize every client/server implementation to adhere to their differences from the core specification.

> Whatever is `@required` is always meant to be filled with a meaningful value, and for smithy to prescribe that a `@required` => `@default` modification is always non-breaking, whilst dictating that it's okay for `@default` fields to sometimes not appear over the wire, would mean that it's okay for services to start breaking the contractual promise to always fill a meaningful value there.

I disagree that this would be a contractual break in Smithy, but I can see how a specific protocol might consider it a break in the serialization layer. The current wording says default values SHOULD NOT be serialized. I think that's a safe default stance to take for protocol implementations, and because it's a SHOULD, future protocols are free to deviate if necessary. As I said in my previous reply, I'll take a pass at the proposal to clear this up.

> Smithy dictating this (regardless of protocols) also remove organisations ability to define protocols where the absence of value and the presence of a default value (in the serialised messages) have the same meaning. This might not occur often, and you might think that the difference is meaningless in most usecases you think about, but I certainly think this should be discussed at the level of protocols.

Agreed that whether a default value is serialized or not isn't a hard requirement of this proposal.

Whether implementations can differentiate between explicit default values and default values *is* something this proposal and Smithy in general should standardize because it affects the compatibility of tools across implementations and programming languages.

> Also I kinda dislike the wording "we typically don't serialize null values in Smithy", when I think you mean "we typically don't serialise null values in the AWS-specific protocols we defined using Smithy". I think the distinction is important to be made if you want smithy to outlive the boundaries of AWS and become an industry standard ... which I very much think it should !

That's a fair point, and I'll keep that in mind. I often think about AWS because we have tens of thousands of models defined in Smithy or Smithy's internal predecessor, which gives a huge corpus of information for me to pull from, and lessons learned. I do think that the guidance around information disclosure is super useful to document, and teams should care about it, but you're right that Smithy != AWS.

Reply:
Hi Michael, and thanks for the detailed reply, this is much appreciated. 

> I mention in the motivation section that nullability is equivalent to optionality, but would a more tactical solution here be to move that guidance into the intro?

This has probably eluded me, and I think it should be the case. 

>  The usage of the @default trait forfeits the knowledge of whether a type is set explicitly to a zero value or defaulted to a zero value. If that information is required, then @default / @required shouldn't be used. For this use case you need to use nullable/optional/maybe/whatever you wanna call it members, which provides the member the ability to differentiate between set and unset.

I can personally agree with that statement. I can also say that I know a great many people that will have strong opinions about it (we Scala devs tend to be extremely opinionated about everything, as you may already have guessed 😅)

Anyhow, regardless of the MUST vs SHOULD debate, as long as it's made clear the expectations around how `@default` values are serialised (or absent from the serialisation) are protocol-specific, my main concern is mostly addressed ... except maybe for the case of structures / unions.  

#### Regarding structures and unions 

Somewhere else in the document, you state the following (quoting here to avoid splitting the convo) 

> The following list describes the default zero value of each kind of shape. Programming languages and code generators that cannot initialize structure members with the above default values MUST continue to represent those members as nullable.

and

> Structure: no zero value. Structures are always nullable and never have a
  default value. In order for a structure to have a default value, all members
  of the structure would need to be optional. Because it is a backward
  compatible change to remove the `@required` trait from a structure member,
  whether a structure has a zero value or not is too volatile to rely on for
  code generation.
> Union: no zero value. Unions are always nullable, MUST be set to one of its
  variants for it to be valid, and have no default variant. Adding some kind of
  special "default" or "unset" value to unions in code to account for default
  values would ultimately require _more_ special-case code in generated code to
  use than just dealing with nullable unions.

I do not see this as universally correct. This is making @required inherently useless on structures and unions, and implies that client code has to protect against nullity no matter what, as they cannot trust that the service will keep sending the data they rely on. It's making life a lot more convenient for the service (by relaxing its ability to evolve by a LOT) and a lot more inconvenient to the client. It drastically impacts developer experience on both sides, by forcing client users to over protect (which is like, THE thing that Scala devs find the most annoying about using the java AWS SDKs), and preventing server users from trusting the type systems to guide them through what it they should set and what they shouldn't set. 

I also find it paradoxal that the removal of `@required` on members targeting structures would be considered safe wrt compatibility, given the current [specs](https://awslabs.github.io/smithy/1.0/spec/core/constraint-traits.html?highlight=required#required-trait). The wording on the specs imply that clients should be able to trust that a required value will always appear in the output of operations. Making it okay for `@required` to be removed from output shapes would imply that clients cannot trust that information. 

#### Playing with an open hand 

I can tell you right now that as a third party, I will not abide by the directive to represent all structures/union fields as optional in the tooling that I'm building : in the generated Scala code, members targeting structures and unions and that have `@required` will be generated in an unwrapped fashion, whereas those who don't will be generated wrapped in `Option`. This is just more idiomatic, and better UX for Scala developers, in particular in the usecase that I'm setting up in my organisation. I might decide to bake some opt-in logic to abide by your directive in my code generator, so that in the context of a pure-Scala AWS SDK, clients would not start breaking when AWS performs what is, in my opinion, a breaking change (ie the removal of `@required` on field that appears in covariant/output position). 

I can tell you however that I am also building some tooling using `smithy-diff` to check for breaking changes, without relying on the default validation events, but rather implementing my own events, abiding by some rules of compatibility that I can prove formally (using a proof system). This tooling will be used internally in my organisation. 

I am **extremely appreciative** with the fact that he tooling empowers me to do so, and even though I disagree with some of aspects of the specification, as long as I'm able to apply discretion and that the smithy libraries do not try to overly protect against these usecases, I'll continue championing the use of smithy in my organisation. 

#### In conclusion 

I think we have fundamentally diverging opinions on the guarantees that users should be able to expect when using `@required`. The smithy libraries (`smithy-model`, `smithy-diff`) are, I think, very well designed and as long as they are flexible enough to cater to my organisation needs, I think the divergence of opinion isn't too problematic. In particular : as long as the use of `@required` does not become more constrained than it already is (ie, the smithy team might decide that it's not possible to put @required on members targeting structures/unions anymore), I'm happy to take responsibility for whatever pain points might come from me making the decision to deviate from the specs. 

Overall, I'm am in agreement with most of the specification as it exists, and I am partially in agreement with the `default` proposal. 

I do not have expectations for the smithy team to take steps towards something that would cater to my opinions/usecases, but if it was the case, a possible course of action would be for another trait to be introduced, with stricter semantics than `@required` (could call it `strictRequired` for instance) 

Kind regards,